// ra-core/dist/cjs/util/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMutationMode = exports.useWhyDidYouUpdate = exports.warning = exports.removeKey = exports.removeEmpty = exports.Ready = exports.ComponentPropType = exports.getFieldLabelTranslationArgs = exports.FieldTitle = exports.escapePath = void 0;
var escapePath_1 = __importDefault(require("./escapePath"));
exports.escapePath = escapePath_1.default;
var FieldTitle_1 = __importDefault(require("./FieldTitle"));
exports.FieldTitle = FieldTitle_1.default;
var getFieldLabelTranslationArgs_1 = __importDefault(require("./getFieldLabelTranslationArgs"));
exports.getFieldLabelTranslationArgs = getFieldLabelTranslationArgs_1.default;
var ComponentPropType_1 = __importDefault(require("./ComponentPropType"));
exports.ComponentPropType = ComponentPropType_1.default;
var removeEmpty_1 = __importDefault(require("./removeEmpty"));
exports.removeEmpty = removeEmpty_1.default;
var removeKey_1 = __importDefault(require("./removeKey"));
exports.removeKey = removeKey_1.default;
var Ready_1 = __importDefault(require("./Ready"));
exports.Ready = Ready_1.default;
var warning_1 = __importDefault(require("./warning"));
exports.warning = warning_1.default;
var useWhyDidYouUpdate_1 = __importDefault(require("./useWhyDidYouUpdate"));
exports.useWhyDidYouUpdate = useWhyDidYouUpdate_1.default;
var getMutationMode_1 = require("./getMutationMode");
Object.defineProperty(exports, "getMutationMode", { enumerable: true, get: function () { return getMutationMode_1.getMutationMode; } });
__exportStar(require("./mergeRefs"), exports);
__exportStar(require("./useEvent"), exports);
__exportStar(require("./hooks"), exports);
__exportStar(require("./shallowEqual"), exports);
__exportStar(require("./LabelPrefixContext"), exports);
__exportStar(require("./LabelPrefixContextProvider"), exports);
__exportStar(require("./useLabelPrefix"), exports);
