// ra-core/dist/cjs/preferences/PreferencesEditorContextProvider.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PreferencesEditorContextProvider = void 0;
var React = __importStar(require("react"));
var react_1 = require("react");
var PreferencesEditorContext_1 = require("./PreferencesEditorContext");
var PreferencesEditorContextProvider = function (_a) {
    var children = _a.children;
    var _b = (0, react_1.useState)(false), isEnabled = _b[0], setIsEnabled = _b[1];
    var _c = (0, react_1.useState)(null), editor = _c[0], setEditor = _c[1];
    var _d = (0, react_1.useState)(), preferenceKey = _d[0], setPreferenceKey = _d[1];
    var _e = (0, react_1.useState)(null), path = _e[0], setPath = _e[1];
    var _f = (0, react_1.useState)(null), title = _f[0], setTitleString = _f[1];
    var _g = (0, react_1.useState)(), titleOptions = _g[0], setTitleOptions = _g[1];
    var enable = (0, react_1.useCallback)(function () { return setIsEnabled(true); }, []);
    var disable = (0, react_1.useCallback)(function () {
        setIsEnabled(false);
        setEditor(null);
    }, []);
    var setTitle = (0, react_1.useCallback)(function (title, titleOptions) {
        setTitleString(title);
        setTitleOptions(titleOptions);
    }, []);
    var context = (0, react_1.useMemo)(function () {
        return {
            editor: editor,
            setEditor: setEditor,
            preferenceKey: preferenceKey,
            setPreferenceKey: setPreferenceKey,
            title: title,
            titleOptions: titleOptions,
            setTitle: setTitle,
            isEnabled: isEnabled,
            disable: disable,
            enable: enable,
            path: path,
            setPath: setPath,
        };
    }, [
        disable,
        enable,
        editor,
        preferenceKey,
        isEnabled,
        path,
        setPath,
        title,
        titleOptions,
        setTitle,
    ]);
    return (React.createElement(PreferencesEditorContext_1.PreferencesEditorContext.Provider, { value: context }, children));
};
exports.PreferencesEditorContextProvider = PreferencesEditorContextProvider;
