// ra-ui-materialui/dist/cjs/layout/DashboardMenuItem.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DashboardMenuItem = void 0;
var react_1 = __importDefault(require("react"));
var prop_types_1 = __importDefault(require("prop-types"));
var Dashboard_1 = __importDefault(require("@mui/icons-material/Dashboard"));
var ra_core_1 = require("ra-core");
var MenuItemLink_1 = require("./MenuItemLink");
var DashboardMenuItem = function (props) {
    var basename = (0, ra_core_1.useBasename)();
    var _a = props.leftIcon, leftIcon = _a === void 0 ? react_1.default.createElement(Dashboard_1.default, null) : _a, _b = props.to, to = _b === void 0 ? "".concat(basename, "/") : _b, _c = props.primaryText, primaryText = _c === void 0 ? 'ra.page.dashboard' : _c, rest = __rest(props, ["leftIcon", "to", "primaryText"]);
    return (react_1.default.createElement(MenuItemLink_1.MenuItemLink, __assign({ leftIcon: leftIcon, to: to, primaryText: primaryText }, rest)));
};
exports.DashboardMenuItem = DashboardMenuItem;
exports.DashboardMenuItem.propTypes = {
    leftIcon: prop_types_1.default.element,
    locale: prop_types_1.default.string,
    onClick: prop_types_1.default.func,
    dense: prop_types_1.default.bool,
    sidebarIsOpen: prop_types_1.default.bool,
};
